### ISSUE TITLE
ArchSpace - Wrong computed dimensions (floor area and perimeter length) with non-prismatic rooms

### ISSUE THREAD

**Issue by marcuspollio** [FreeCAD/FreeCAD#13424]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

The ArchSpace Object features floor area and perimeter length properties.
These are computed at the center of mass, according to the [wwiki documentation](https://wiki.freecad.org/Arch_Space) and L:447-491 of ArchSpace.py  

https://github.com/FreeCAD/FreeCAD/blob/afaf0ce8ea2d625e0d624374e1836a874b571714/src/Mod/Arch/ArchSpace.py#L447-L491

However, there are three problems:

- This method is not suitable for all room shapes, especially those with variable heights (such as under the roof) or those with curved aspects (e.g. arches, domes, vaults). Partially linked to issue #6091.

- The footprint, as required by [IfcSpace](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSpace.htm), is not exposed. In a second stage, the IFC scheme suggests access to other [useful element quantities](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Qto_SpaceBaseQuantities.htm).

- Depending on the code, norm or client, the space areas must be calculated at a certain height (e.g. 1.8m or 1.5m). Currently, there is no direct way to expose that.

To reproduce:
- Launch FreeCAD, in the Start Page, choose BIM/Architecture
- Create any non-prismatic object (any way you like, Draft, Sketcher->Pad, etc.). See below for example.
- Turn it into an ArchSpace object (SP shortcut). Or create an ArchSpace from any other suitable object boundaries.
- In the Data tab of the Properties panel, check the Perimeter Length (in Component section) and the floor Area (in the Space section) values.

![ArchSpaceBug](https://github.com/FreeCAD/FreeCAD/assets/131592747/ee5a75be-9a91-4f6e-84b4-d45bc80fda33)

In this simple example, the Space has a length of 5m, depth of 1m, a lower height of 1m and upper height of 4m.
The footprint is 5m². The perimeter is 12m. The center of mass is at a height of 1.25m but not very useful in this case.

Three possible solutions to discuss :

1) Update the terminology and tooltips to indicate what is really computed : an horizontal section at the center of mass and not a "floor area".

2) Use the footprint area (IFC method) and use this terminology instead to indicate it's a horizontal projection of the whole shape.

3) Provide a greater choice of element quantities to refer from (such as the ones listed on Ifc SpaceBaseQuantities).

### Full version info

```shell
OS: GNU/Linux Ubuntu 22.04 LTS based (DE:GNOME/pop)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36811 (Git)
Build type: Release
Branch: main
Hash: c0d8de527adb9d68716df092e2fff497ae29d5ea
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.5.1
```

### Subproject(s) affected?

Arch

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by yorikvanhavre** [comment]
Indeed these measurements are currently taken at the center of mass of the space volume. And indeed that's more than often not what we want. the problem is how to determine exactly the surface we want to take the perimeter and area from. We could take the face with the lower Z coordinate. But what if the space's "ground face" is made of several faces? What if there is a level change (more than one face with different Z)?

That was the advantage of the current method, it doesn't matter how the floor looks like.

I have no clear idea on the best way to approach this. Maybe offer different strategies? Or allow the user to pick the ground face(s) manually?

**Comment by marcuspollio** [comment]
@yorikvanhavre Thank you for having a look.

What about the three possible solutions I provided in OP ?

**Comment by yorikvanhavre** [comment]
Hmm indeed I hadn't looked well :) Projecting is interesting indeed! I suggest maybe the following:

* The ArchSpace gains a new property, ex: **AreaCalculation** which is a drop-down list with (currently) two values: **Center of Mass** and ** Z Projection** (better ideas?). That property defines how the area and perimeter is calculated

* The Center of Mass method does exactly what it does now

* The Z Projection flattens the space shape onto the XY plane (the [ArchVRM](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchVRM.py) module can do that, the faces then need to be unioned, the areas summed and the perimeter edges extracted)

* Further methods can be easily added, I'd for ex. imagine we could add both 1.8m and 1.5m, which would 1) calculate the lowest Z value, 2) find a point 1.8m (or 1.5m or whatever) above that, and cut the volume at that point

How does that sound?

**Comment by DeniseBryson** [comment]
Hi,
I am new. I wanted to start with my first Issue and found this discussion. :)

On my search where the PerimeterLength is set. 
I found two function which according to the comments seem to compute the wanted parameters.
The first one is in [ArchComponent.py](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchComponent.py#L963)
It seems like the result of this function is overwritten by the getArea function in [ArchSpace.py](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchSpace.py#L434)


The other function is in [ArchSite.py](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchSite.py#L739)
I guess that is a different type.

**Comment by marcuspollio** [comment]
Welcome @DeniseBryson 

Let us know if you submit a Pull Request so we can test it and provide feedback.
Thank you !

**Comment by yorikvanhavre** [comment]
@DeniseBryson there are many functions all over the FreeCAD code that provide perimeters, but here I'm afraid none of those will be usable because what we are trying to do here is different from these other cases.

Basically, if you want to follow the plan I sketched above, I would start with this:

* Add a new App::PropertyEnum to the Arch_Space object that allows to switch between the two modes, At center and projected
* Rename the getFootprint to something like getFootprintCenter
* Implement another one, something like getFootPrintProjected
* Inside that new function, you'll want to do this:
   1. Get a list of all the faces
   2. Project all the faces to the XY plane
   3. Union (fuse) all these faces
   4. Get the area
   5. Get the border edges (basically edges that are used only once)
   6. Turn these edges into wires
   7. Get the length of the outer wire (verify if there are holes)
* Make a new getFootprint, and inside, use getFootprintCenter or getFootprintProjected depending on the value of our Enum property

I'll let you explore this, feel free to ask for help if you're lost!

**Comment by DeniseBryson** [comment]
Thanks a lot for the advice, I will try that!

**Comment by DeniseBryson** [comment]
Hi again,
so I found that in the example above, the object has a property called HorizontalArea, that is computed by the area of the  XY-projections of all faces. I do not understand why the GUI does not show this parameter in the component section but you can see it in the python console with:
`obj.HorizontalArea`
It has a small rounding error though: HorizontalArea = 4999999.99mm^2 and not HorizontalArea = 5m^2
The Perimeter is also set according to this area, but is later overwritten by the computeArea method of the Space object.

My suggestion would be to make the HorizontalArea visible in the component section, do you have any hints why it is not showing right now. 
Eliminate the rounding error, do you have an idea how this could be done?

And to create an extra pair of properties in the Space section called PerimeterCoM and AreaCoM, which are computed by the current function, computeArea cited at the top.

What do you think?

**Comment by yorikvanhavre** [comment]
Oh yes, I had forgotten about that indeed. All Arch objects have a HorizontalArea which is indeed exactly what we want to do, and Spaces are also Arch objects...

So our task here is dramatically simplified:

* Create an enum property to switch area calculation
* If projected area is selected, simply copy the value of HorizontalArea to Area
* In all cases, Indeed a Perimeter property would be interesting to add

There is the problem that HorizontalArea is calculated at ArchComponent level, while the Area (when taken at center of mass) at the ArchSpace level, which makes it complicated to add a function that calculates the perimeter in both cases. Maybe what we could do here, is make those two functions save the faces they produced (in self.horizontalfaces or something like that), then in ArchSpace use that to calculate the perimeter, for ex. using https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Draft/draftgeoutils/faces.py#L61

**Comment by DeniseBryson** [comment]
Hi,
sorry to make this so complicated, but I stumbled on something else now.
The function ArchComponent.computeAreas, that computes the horizontal area, also calculates vertical areas and the perimeter of the projection of all faces onto the xy-plane. 
The vertical areas are calculated correctly, but the Perimeter is not.

There are two problems here:

- First the code indicates that horizontal faces should not be projected at all, but in the current example the horizontal face is just left out.

    This can be fixed by:
    - Considering all faces that have a different angle than 1.57 [here](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchComponent.py#L1018)
    - Accepting angles equal to Pi as well instead of only accepting angles equal to 0 with respect to the unit vector (0,0,1). [here](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchComponent.py#L1027)
    
    Maybe this whole function "computeAreas" should be refactored a little, to avoid calling the getangle several times. and to eliminate the inaccuracies..

- Secondly, after the projection of the non horizontal area in the example the perimeter is 11m not 12m.
I do not understand this problem the area is computed correctly but the perimeter afterwards is 11m.
If you project the horizontal area the perimeter also changes to 11m. Am i missing something here?
[Here](https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Arch/ArchComponent.py#L1032) is the function that causes this behavior.

Please tell me if this is going to far, but I feel like fixing the initial problem requires fixes in ArchComponent.computeAreas as well. The properties horizontal area and its perimeter should not be calculated twice.

**Comment by yorikvanhavre** [comment]
Problem 1: the horizontal faces are appended (L1029). They just don't need to be projected, because they are horizontal already.

Problem 2:  I think the problem here is 1) since the horizontal faces are not projected, they could be not on the XY plane, but ex. above. So they won't fuse nicely with the others and 2) the removesplitter() function often fails, so there is still more than one face in the set.

So indeed we need to fix the perimeter calculation. I would to this:

1. Project the horizontal faces too. So the projected face will lie on the XY plane too and will fuse correctly with the others. Or at least, not project those that are horizontal AND their CenterOfMass.z == 0. But the current code also does not consider that a surface might not be flat, so in any case it's best to project everything
2. Instead of relying on the OuterWire to get the perimeter, which won't work in case there is more than one face, build wires from all edges that are shared by only one face (which are the border edges), something like this:

```python
edges_table = {}
for e in shape.Edges:
    edges_table.setdefault(e.hashCode(),[]).append(e)
border_edges = [pair[0] for pair in edges_table.values() if len(pair) == 1]
perimeter = sum([e.Length for e in border_edges])
```

**Comment by DeniseBryson** [comment]
Thanks for the advice! I implemented this.
But there is another problem now...
The protection function returns faces with one edge missing. This explains the perimeter of 11m instead of 12m. One 1m edge just disappeared after the projection. I do not know how to fix that. Is there a function that redraws all boarder edges of a face?

**Comment by yorikvanhavre** [comment]
That's weird.. Which function is that?

**Comment by DeniseBryson** [comment]
> That's weird.. Which function is that?

`Part.Face(Part.Wire(TechDraw.project(f,FreeCAD.Vector(0,0,1))[0].Edges))`

in line 1032

**Comment by DeniseBryson** [comment]
If you use this line with the top or floor face of the example above, both resulting faces have only 3 edges.
the vertical faces have two edges after the protection.
my problem is that i can not find the function in src/Mod/TechDraw.
if I know where it is I can search for reason of this behaviour. 
should I open another issue for this?

**Comment by yorikvanhavre** [comment]
The projection algorithms all come from opencascade, I don't think we can easily fix this here.. But did you check if `TechDraw.project(f,FreeCAD.Vector(0,0,1))[0].Edges` returns only 3 edges? The problem might be in the Part.Wire() or Part.Face() steps, which might be easier to fix.

**Comment by DeniseBryson** [comment]
True! It's Part.Wire() that drops one edge.
I try to look into that.

**Comment by yorikvanhavre** [comment]
Ah then we have a good way forward. Part.Wire is often failing, and it's not always clear why. We should use instead `DraftGeomUtils.findWires(edges)`. That function will return a list of wires (it could detect more than one loop).

I think that might solve our problem

**Comment by DeniseBryson** [comment]
Perfect that works!

Unfortunately I have to more problems.
How do I detect the event when the users is switching the calculation type in the enum field? 

And it seems like another display error got introduced, but I am not sure if i caused it. I did not change anything regarding "Elevation With Flooring". Do you have any idea what could cause this?

![grafik](https://github.com/FreeCAD/FreeCAD/assets/18719132/64487a2c-7f3c-4b49-a3b9-884dd4b19b17)

**Comment by yorikvanhavre** [comment]
Cool! No worries, we explain you on the way :)

> How do I detect the event when the users is switching the calculation type in the enum field?

Basically when  a property changes, two things happen:

1) onChanged(obj, prop) is called. There you have a first chance to handle specifically that property.
2) execute(obj) is called next. 

There, usually the shape is recomputed. In our case, the mechanism that recomputes the area is called every time the execute() function is run. which is good because we basically want to recompute the area whenever the shape is changed.

So it depends on what you want to do here. I would suggest just, inside the area computation code, see if the enum property is one or the other, and use one function or another depending on that,, to calculate the area.

**Comment by DeniseBryson** [comment]
thanks again for the help!
I opened a [pull request](https://github.com/FreeCAD/FreeCAD/pull/14211), where I tried to in-cooperate all your comments.
